File: backend/app/routers/accounts.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.account import AccountCreate, AccountRead, AccountUpdate
from app.services.account_service import (
    create_account,
    delete_account,
    get_account,
    get_accounts,
    update_account,
)

router = APIRouter(prefix="/accounts", tags=["accounts"])


@router.post("/", response_model=AccountRead, status_code=status.HTTP_201_CREATED)
async def create_account_endpoint(
    data: AccountCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    account = await create_account(db, current_user.id, data)
    return account


@router.get("/", response_model=list[AccountRead])
async def list_accounts(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await get_accounts(db, current_user.id)


@router.get("/{account_id}", response_model=AccountRead)
async def get_account_endpoint(
    account_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    account = await get_account(db, current_user.id, account_id)
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    return account


@router.patch("/{account_id}", response_model=AccountRead)
async def update_account_endpoint(
    account_id: uuid.UUID,
    data: AccountUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    account = await update_account(db, current_user.id, account_id, data)
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    return account


@router.delete("/{account_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_account_endpoint(
    account_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    deleted = await delete_account(db, current_user.id, account_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Account not found")

File: frontend/src/api/accounts.ts
import { apiFetch } from "@/api/client";
import type { Account, CreateAccountData } from "@/types/models";

export async function getAccounts(token: string): Promise<Account[]> {
  return apiFetch<Account[]>("/api/v1/accounts", {}, token);
}

export async function createAccount(
  data: CreateAccountData,
  token: string
): Promise<Account> {
  return apiFetch<Account>(
    "/api/v1/accounts",
    {
      method: "POST",
      body: JSON.stringify(data),
    },
    token
  );
}

export async function updateAccount(
  accountId: string,
  data: Partial<CreateAccountData>,
  token: string
): Promise<Account> {
  return apiFetch<Account>(
    `/api/v1/accounts/${accountId}`,
    {
      method: "PATCH",
      body: JSON.stringify(data),
    },
    token
  );
}
